store/data_source.go

```diff
@@ -536,3 +554,188 @@ func (*Store) listDataSourceV2(ctx context.Context, tx *Tx, instanceID string) (
 
 	return dataSourceMessages, nil
 }
+
+// AddDataSourceToInstanceV2 adds a RO data source to an instance and return the instance where the data source is added.
+func (s *Store) AddDataSourceToInstanceV2(ctx context.Context, instanceUID, creatorID int, dataSource *DataSourceMessage) (*InstanceMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, errors.New("Failed to begin transaction")
+	}
+	defer tx.Rollback()
+
+	instance, err := s.addDataSourceToInstanceImplV2(ctx, tx, instanceUID, creatorID, dataSource)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return nil, errors.New("Failed to commit transaction")
+	}
+
+	s.instanceCache.Store(getInstanceCacheKey(instance.EnvironmentID, instance.ResourceID), instance)
+	s.instanceIDCache.Store(instance.UID, instance)
+	return instance, nil
+}
+
+// RemoveDataSourceV2 removes a RO data source from an instance and return the instance where the data source had beed removed from.
+func (s *Store) RemoveDataSourceV2(ctx context.Context, instanceUID int, dataSourceTp api.DataSourceType) (*InstanceMessage, error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, errors.New("Failed to begin transaction")
+	}
+	defer tx.Rollback()
+
+	result, err := tx.ExecContext(ctx, `
+		DELETE FROM data_source WHERE instance_id = $1 AND type = $2;
+	`, instanceUID, dataSourceTp)
+	if err != nil {
+		return nil, FormatError(err)
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get rows affected")
+	}
+	if rowsAffected != 1 {
+		return nil, errors.Errorf("deleted %v read-only data source records from instance %v, but expected one", rowsAffected, instanceUID)
+	}
+
+	instance, err := s.findInstanceImplV2(ctx, tx, &FindInstanceMessage{
```

Comment:
Instead of having a find with txn resulting conflicts and complexity, let's just commit and use GetInstanceV2 for returning the instance. On the other hand, maybe this store function doesn't need to do that, the caller can get instance and return.

Reply:
Done.